homeassistant/components/dreo/fan.py

```diff
@@ -0,0 +1,145 @@
+from homeassistant.util.percentage import (
+    percentage_to_ranged_value,
+    ranged_value_to_percentage, int_states_in_range
+)
+from hscloud.const import DEVICE_TYPE, FAN_DEVICE
+
+from . import DreoConfigEntry
+from .entity import DreoEntity
+from hscloud.hscloudexception import HsCloudException, HsCloudBusinessException
+
+_LOGGER = logging.getLogger(__name__)
+
+async def async_setup_entry(
+        hass: HomeAssistant,
+        config_entry: DreoConfigEntry,
+        async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Fan from a config entry."""
+
+    async_add_entities([
+        DreoFanHA(device, config_entry)
+        for device in config_entry.runtime_data.devices
+        if DEVICE_TYPE.get(device.get("model")) == FAN_DEVICE.get("type")
+    ])
+
+
+class DreoFanHA(DreoEntity, FanEntity):
+    """Dreo fan."""
+
+    _attr_supported_features = (FanEntityFeature.PRESET_MODE
+                                | FanEntityFeature.SET_SPEED
+                                | FanEntityFeature.OSCILLATE)
+
+    def __init__(self, device, config_entry) -> None:
+        """Initialize the Dreo fan."""
+        super().__init__(device, config_entry)
+        self._attr_name = None
+        self._attr_preset_modes = FAN_DEVICE.get("config").get(self._model).get("preset_modes")
+        self._attr_low_high_range = FAN_DEVICE.get("config").get(self._model).get("speed_range")
+        self._attr_speed_count = int_states_in_range(self._attr_low_high_range)
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return True if device is on."""
+        return self._attr_state
+
+    @cached_property
+    def preset_modes(self) -> list[str] | None:
+        """Return a list of available preset modes.
+
+        Requires FanEntityFeature.SET_SPEED.
+        """
+        if hasattr(self, "_attr_preset_modes"):
+            return self._attr_preset_modes
+        return None
+
+    @cached_property
+    def speed_count(self) -> int:
+        """Return the number of speeds the fan supports."""
+        if hasattr(self, "_attr_speed_count"):
+            return self._attr_speed_count
+        return 6
+
+    def turn_on(self, **kwargs: Any) -> None:
+        """Turn the device on."""
+        self._try_command(
+            "Turn the device on failed.", power_switch=True
+        )
+        self._attr_state = True
+
+    def turn_off(self, **kwargs: Any) -> None:
+        """Turn the device off."""
+        self._try_command(
+            "Turn the device off failed.", power_switch=False
+        )
+        self._attr_state = False
+
+    def set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of fan."""
+        self._try_command(
+            "Set the preset mode failed.", mode=preset_mode
+        )
+        self._attr_preset_mode = preset_mode
+
+    def set_percentage(self, percentage: int) -> None:
+        """Set the speed of fan."""
+        if percentage > 0:
```

Comment:
Is the fan somehow already turned off in case percentage is 0 or does it make sense to have an else case here?

Reply:
When the percentage is 0, home-assistant will call turn_off to turn off the fan, so calling self.turn_off() again does not make sense.